ibis/common/grounds.py

```diff
@@ -0,0 +1,246 @@
+from __future__ import annotations
+
+import inspect
+from abc import ABC, ABCMeta, abstractmethod
+from typing import Any, Hashable
+from weakref import WeakValueDictionary
+
+from ibis.util import frozendict
+
+from .. import util
+from .caching import WeakCache
+from .validators import Optional, Validator
+
+EMPTY = inspect.Parameter.empty  # marker for missing argument
+
+
+class BaseMeta(ABCMeta):
+
+    __slots__ = ()
+
+    def __call__(cls, *args, **kwargs):
+        return cls.__create__(*args, **kwargs)
+
+
+class Base(metaclass=BaseMeta):
+
+    __slots__ = ()
+
+    @classmethod
+    def __create__(cls, *args, **kwargs):
+        return type.__call__(cls, *args, **kwargs)
+
+
+class Parameter(inspect.Parameter):
+    """
+    Augmented Parameter class to additionally hold a validator object.
+    """
+
+    __slots__ = ('_validator',)
+
+    def __init__(
+        self,
+        name,
+        kind=inspect.Parameter.POSITIONAL_OR_KEYWORD,
+        *,
+        validator=EMPTY,
+    ):
+        super().__init__(
+            name,
+            kind,
+            default=None if isinstance(validator, Optional) else EMPTY,
+        )
+        self._validator = validator
+
+    @property
+    def validator(self):
+        return self._validator
+
+    def validate(self, this, arg):
+        if self.validator is EMPTY:
+            return arg
+        else:
+            # TODO(kszucs): use self._validator
+            return self.validator(arg, this=this)
+
+
+class AnnotableMeta(BaseMeta):
+    """
+    Metaclass to turn class annotations into a validatable function signature.
+    """
+
+    __slots__ = ()
+
+    def __new__(metacls, clsname, bases, dct):
+        # inherit from parent signatures
+        params = {}
+        for parent in bases:
+            try:
+                signature = parent.__signature__
+            except AttributeError:
+                pass
+            else:
+                params.update(signature.parameters)
+        # store the originally inherited keys so we can reorder later
+        inherited = set(params.keys())
```

Comment:
It should be fine to leave this as `params.keys()`.

Reply:
Not sure since we alter the params mapping below.

Reply:
Ah, right. Thanks!